### ISSUE TITLE
Incorrect comma-dangle error shown when using object rest/spread. 

### ISSUE THREAD

**Issue by jakewtaylor** [eslint/eslint#9541]
* **ESLint Version:**
4.3.0
* **Node Version:**
6.1.0.0
* **npm Version:**
3.10.10
**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
    "env": {
        "es6": true,
        "browser": true,
    },
    "extends": "airbnb",
    "plugins": [],
    "parser": "babel-eslint",
    "parserOptions": {
        "ecmaFeatures": {
            "experimentalDecorators": true
        }
    },
    "globals": {
        "SyntheticInputEvent": true,
        "SyntheticKeyboardEvent": true,
    },
    "rules": {
        "indent": ["error", 4, {
            "SwitchCase": 1
        }],
        "space-before-function-paren": ["error", "always"],
        "react/jsx-filename-extension": [1, {
            "extensions": [".js", ".jsx"]
        }],
        "react/jsx-indent": [2, 4],
        "react/jsx-indent-props": [2, 4],
        "import/no-unresolved": [2, {
            "ignore": ['^rdx', '^app', '^utils']
        }],
        "react/sort-comp": [0],
        "import/no-extraneous-dependencies": [0],
        "import/prefer-default-export": [0],
        "import/extensions": [0],
        "no-return-assign": [2, "except-parens"],
        "no-unneeded-ternary": ["error", {
            "defaultAssignment": true
        }],
        "valid-typeof": [0],
        "jsx-a11y/href-no-hash": "off",
        "jsx-a11y/label-has-for": "off",
        "class-methods-use-this": "off",
        "no-class-assign": "off",
        "linebreak-style": ["warn", "unix"]
    }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const mapStateToProps = (
    state,
    {
        id,
        type,
        ...post, // <- ESLint says this comma shouldn't be here
    {,
) => ({
    liked: true,
    ...post,
});
```

```bash
./node_modules/.bin/eslint path/to/file.js
```

**What did you expect to happen?**
No errors.

**What actually happened? Please include the actual, raw output from ESLint.**
```
6:16  error  Unexpected trailing comma  comma-dangle

1 problem (1, error, 0 warnings)
1 error, 0 warnings potentially fixable with the `--fix` option.
```

**Comment by jakewtaylor** [comment]
Sorry, found #7463. Maybe this should be documented though?

**Comment by not-an-aardvark** [comment]
This is caused by incorrect parser behavior (`babel-eslint` incorrectly considers your code to be valid rather than throwing a parser error), so it's not really defined behavior for the `comma-dangle` rule.

**Comment by rbalicki2** [comment]
Could there please be an option to allow this? Perhaps an option like `objects: ['always-multiline', { incorrectlyTolerateRestOperatorComma: true }]` ;)

**Comment by not-an-aardvark** [comment]
Could you elaborate on why you would want an option like that? As mentioned in #7463, trailing commas after rest operators are syntactically invalid. I don't see how it would be useful to have the rule allow them.

**Comment by rbalicki2** [comment]
It's a purely stylistic thing, and of course, I understand all the arguments to the contrary. It's simply that the comma looks good and consistent, and seems logical. Visual consistently is important because when attempting to add lint rules to an existing code base and modify others' behaviors, forcing developers to think causes them to resist the lint rules :)

```
// e.g. the comma looks good here:
const MyComponent = ({
  className,
  ...props,
}) => { ... }
```

**Comment by not-an-aardvark** [comment]
Sure, but I don't think it makes sense to use invalid JavaScript instead of valid JavaScript just for visual consistency. Presumably, you will need to run your code at some point, and JavaScript engines can only run valid JavaScript code.

I don't think it would make sense to add an option that would encourage users to write invalid JavaScript. Of course, you could create a plugin that does this if you'd like, so that you could enforce it on an existing codebase (provided that you're using a parser that also handles invalid trailing commas.)
